read.go

```diff
@@ -217,57 +217,62 @@ func (c *Conn) readLoop(ctx context.Context) (header, error) {
 	}
 }
 
-func (c *Conn) readFrameHeader(ctx context.Context) (header, error) {
+func (c *Conn) prepareRead(ctx context.Context, err *error) (func(), error) {
 	select {
 	case <-c.closed:
-		return header{}, net.ErrClosed
+		return nil, net.ErrClosed
 	case c.readTimeout <- ctx:
 	}
 
-	h, err := readFrameHeader(c.br, c.readHeaderBuf[:])
-	if err != nil {
+	done := func() {
 		select {
 		case <-c.closed:
-			return header{}, net.ErrClosed
-		case <-ctx.Done():
-			return header{}, ctx.Err()
-		default:
-			return header{}, err
+			if *err != nil {
+				*err = net.ErrClosed
+			}
+		case c.readTimeout <- context.Background():
+		}
+		if *err != nil && ctx.Err() != nil {
+			*err = ctx.Err()
 		}
 	}
 
-	select {
-	case <-c.closed:
-		return header{}, net.ErrClosed
-	case c.readTimeout <- context.Background():
+	c.closeStateMu.Lock()
+	closeReceivedErr := c.closeReceivedErr
+	c.closeStateMu.Unlock()
+	if closeReceivedErr != nil {
+		defer done()
```

Comment:
Shouldn't we make this behavior consistent - in this line you call `done()`, but at L248 you pass it up. Maybe we should refactor to pass it in both cases?

Reply:
It's not typical to return _both_ an error and a valid value (e.g. if there's an error caller assumes done is nil). The reason we defer done here is that we need to clean up the change we made (readTimeout ctx) if we're going to abort.

Reply:
I see, it just seems like a typical `try-finally` block, but in this case `finally` can be executed on different levels. I was wondering if we can make it consistent.

Reply:
I was thinking a similar thing, would it be weird not to return an error?  For the caller, seems fair to assume they must check `err` even if not returned since they are passing it in by reference?

idk if this is any better though, mostly just thinking out loud

```go
defer prepareRead(ctx, &err)()
if err != nil {
  return header{}, err
}
```

Reply:
That's a clever suggestion @code-asher, I feel perhaps a bit too clever for me, though. Personally I like the explicitness of error handling in Go, albeit verbose, it always tells me when I need to handle an error. In the proposed syntax, I feel that err handling is not entirely obvious. I'll just go with what we have, as the only alternative I can think of is to undo the function and repeat the code between the two functions that use it, whichever is better feels like a coin toss. 😄